examples/petsc/bpssphere.c

```diff
@@ -0,0 +1,366 @@
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+//                        libCEED + PETSc Example: CEED BPs
+//
+// This example demonstrates a simple usage of libCEED with PETSc to solve the
+// CEED BP benchmark problems, see http://ceed.exascaleproject.org/bps,
+// on a closed surface, such as the one of a discrete sphere.
+//
+// The code uses higher level communication protocols in DMPlex.
+//
+// Build with:
+//
+//     make bpssphere [PETSC_DIR=</path/to/petsc>] [CEED_DIR=</path/to/libceed>]
+//
+// Sample runs:
+//
+//     bpssphere -problem bp1 -degree 3
+//     bpssphere -problem bp2 -ceed /cpu/self -degree 3
+//     bpssphere -problem bp3 -ceed /cpu/self -degree 3
+//     bpssphere -problem bp4 -ceed /cpu/occa -degree 3
+//     bpssphere -problem bp5 -ceed /cpu/occa -degree 3
+//     bpssphere -problem bp6 -ceed /cpu/self -degree 3
+//
+//TESTARGS -ceed {ceed_resource} -test -problem bp3 -degree 3
+
+/// @file
+/// CEED BPs example using PETSc with DMPlex
+/// See bps.c for a "raw" implementation using a structured grid.
+/// and bpsdmplex.c for an implementation using an unstructured grid.
+static const char help[] = "Solve CEED BPs on a sphere using DMPlex in PETSc\n";
+
+#include <stdbool.h>
+#include <string.h>
+#include <petscksp.h>
+#include <petscdmplex.h>
+#include <ceed.h>
+#include "setupsphere.h"
+
+int main(int argc, char **argv) {
+  PetscInt ierr;
+  MPI_Comm comm;
+  char ceedresource[PETSC_MAX_PATH_LEN] = "/cpu/self",
+                                          filename[PETSC_MAX_PATH_LEN];
+  double my_rt_start, my_rt, rt_min, rt_max;
+  PetscInt degree = 3, qextra, lsize, gsize, topodim = 2, ncompx = 3,
+           ncompu = 1, xlsize;
+  PetscScalar *r;
+  PetscBool test_mode, benchmark_mode, read_mesh,
+            write_solution, simplex;
+  Vec X, Xloc, rhs, rhsloc;
+  Mat matO;
+  KSP ksp;
+  DM  dm;
+  UserO userO;
+  Ceed ceed;
+  CeedData ceeddata;
+  CeedQFunction qf_error;
+  CeedOperator op_error;
+  CeedVector rhsceed, target;
+  bpType bpChoice;
+
+  ierr = PetscInitialize(&argc, &argv, NULL, help);
+  if (ierr) return ierr;
+  comm = PETSC_COMM_WORLD;
+
+  // Read command line options
```

Comment:
I think that we should do another refactoring pass to identify common code relatively soon in the future. There are a few more common pieces across area, areasphere, bps, bpssphere, multigrid, and setup (or in different subsets) that can probably be rorganized to reduce total number of lines and future mantaience burden while increasing readability and user friendliness. We should also break those big setup.h files into separate .c files eventually.